Profile changes to unblock routing

Abby-Wheelis opened the pull request:
We are almost to the point of being able to implement React routing, but first the parts of ProfileSettings that rely on `settingsScope` (angular) must be converted

this includes: `ControlCollectionHelper`, `ControlSyncHelper` , `showLog`, and `showSensed`

Abby-Wheelis commented:
Update to this PR -- the log and sensed pages have been converted, and are shown below.
<img src="https://github.com/e-mission/e-mission-phone/assets/54848919/c2809a56-5481-4f48-a3a8-b4f473406089" width="350"  /> <img src="https://github.com/e-mission/e-mission-phone/assets/54848919/ffd5f267-db67-4ac6-acaa-2651a4636fe3" width="350"  />

^I just noticed the styling on these doesn't match, so I'll fix that now

There are lots of parts that are not internationalized (title of the page, error messages), @shankari do we want to start internationalizing in dev zone while we're here, or leave it in English?

shankari commented:
@Abby-Wheelis I think that it depends on how much effort it would be. At a high level, any OpenPATH developer is going to have to know English since all our developer documentation is in English. However, the developer zone is not only for developers, we sometimes ask participants to use it for debugging (e.g. are you in the "waiting for trip start" or the "start" state?)

So I guess my argument is that it would be good to translate if it is not too much effort. But if it is going to double the number of strings or something, we have higher priorities.

shankari commented:
> ^I just noticed the styling on these doesn't match, so I'll fix that now

Also, the messages are different - the one on the left shows the actual logs, the one on the right is apparently only showing state transitions??!

Abby-Wheelis commented:
> So I guess my argument is that it would be good to translate if it is not too much effort. But if it is going to double the number of strings or something, we have higher priorities.

I'd say it's ~10 or fewer strings, so worth it! We've updated a bunch recently so I also need to check in with the Lao team again soon (using google translate, and can't double check like I can with Spanish).

Abby-Wheelis commented:
>Also, the messages are different - the one on the left shows the actual logs, the one on the right is apparently only showing state transitions??!

The messages being different is ok, I think, since it's two different pages, which is also what we have in production now. Left page is for "check log" and right is for "check sensed". State transitions is one of three options for sensed data, Locations and Motion Types being the others, which you can change with the hamburger menu. -- Locations and Motion Types are blank lists when I check them here, and on my phone that's on production, so I need to keep looking into if this is what's expected.

shankari commented:
> Locations and Motion Types are blank lists when I check them here, and on my phone that's on production, so I need to keep looking into if this is what's expected.

I think that bitrotted sometime in the distant past. They were supposed to show the list of locations and the list of motion activity objects (read from the usercache) to help debugging. I am fine with removing those - I don't remember the last time that I used them.

Abby-Wheelis commented:
I've added the i18n (only ended up being 4 strings) and removed the bit-rotted options on the sensed page. Translations for our other two languages here: [Dev zone keys](https://github.com/e-mission/e-mission-translate/pull/42#top)

Abby-Wheelis commented:
This fix is far from polished, but I think I've unblocked React routing from the profile at this point. `ControlCollectionHelper`, `ControlSyncHelper` , `showLog`, and `showSensed` have all been rewritten into React components. I was also able to remove all references to `settingScope` because the last reliance on that variable was catching "recomputeAppStatus" broadcasts, which are redundant anyways after the introduction of the custom hook for "on resume".

The migration that I was hoping to do of some of the logical code from `ProfileSettings` into the new helpers is held back by the error message `AlertBars`. If I move the functions that show the bars into the helpers, I would have to pass the visibility states as parameters and leave the bars themselves in `ProfileSettings` because of the way React does not allow exporting functions from within a component. This splitting feels clunky, but might be what React wants. I think ideally there would be some way to throw the "alerts" from within the function like we were in angular/ionic but I haven't figured out how that could work yet. There's a few functions that also make use of the angular service "Logger" which isn't available outside of components (judging by the error messages) so that is another blocker to the "logical migration / cleanup" that I was hoping to start here.

Abby-Wheelis commented:
We just talked about some ways to work through this "alert" issue:

- when we're fully in React Native, we can use React Native's alert
- could use an external library to accomplish something like React Native's alert in the current react native web
- could use the global "window.alert" (as a temporary solution)

Abby-Wheelis commented:
>could potentially render the "force sync" button, etc, as a separate component, export it separately, and embed it directly in profile settings

This is what I've done, for now, with force sync. Now I'm realizing "endForceSync" (ends a trip then forces a sync) relies on that function too. I could easily remedy this if I put the rows next to each other. Currently, `forceSync` is outside of devzone and `endForceSync` is inside. Could they both be inside? How often do users usually force a sync?

shankari commented:
> Now I'm realizing "endForceSync" (ends a trip then forces a sync) relies on that function too. I could easily remedy this if I put the rows next to each other. Currently, forceSync is outside of devzone and endForceSync is inside. Could they both be inside? How often do users usually force a sync?

Can you clarify why do they need to be next to each other given that they are separate components?

Having said that, I am OK with rmoving "Force Sync" into the developer zone. The only reason it is outside right now is because one of the early studies, which only ran for a week, wanted to make sure that participants had uploaded all their data, so wanted to "force sync" at the end of the study.

We haven't had thta requirement for a while now, probably because we are so focused on longitudinal data collection. So I think it is fine to change it now.

Abby-Wheelis commented:
>Can you clarify why do they need to be next to each other given that they are separate components?

While they are separate components, `endForceSync` calls `forceSync` and therefore also relies on the same success/error message components as `forceSync`. If it was just a message, I could use "window.alert" and they wouldn't rely on the same component. But one of the things that `forceSync` can show is a "data pending" consent dialog, that requires collecting and handling user input: 

`then(function(syncPending) {
            Logger.log("sync launched = "+syncPending);
            if (syncPending) {
                Logger.log("data is pending, showing confirm dialog");
                setDataPendingVis(true); //consent handling in modal`

By calling `forceSync`, `endForceSync` needs to share all of the same "hidden" UI components with `forceSync`

shankari commented on www/js/control/ProfileSettings.jsx:
```diff
@@ -24,39 +27,20 @@ const ProfileSettings = () => {
-        controlUpdateCompleteListenerRegistered = true;
-    }
-
     //functions that come directly from an Angular service
```
This comment is no longer true?!

Abby-Wheelis replied:
I updated it, thanks for pointing that out!

shankari commented on www/js/control/ProfileSettings.jsx:
```diff
@@ -65,15 +49,18 @@ const ProfileSettings = () => {
+    const [showingSensed, setShowingSensed] = useState(false);
+    const [showingLog, setShowingLog] = useState(false);
+    const [editSync, setEditSync] = useState(false);
+    const [editCollection, setEditCollection] = useState(false);
```
I'm a bit confused about the interaction between `editCollection` and `editCollectionConfig`.
I see that there is an editCollectionConfig which sets `editCollection` to true, which presumably opens the popup for the edit. But then what does `editCollection` do, and how does it interact with the `refreshCollectionSettings` in the `useEffect` below, and in the `control.update.complete event broadcast` from the helper

Abby-Wheelis replied:
The `control.update.complete` event broadcast cannot be relied on anymore, as it was intertwined with Angular routing. 

`editCollectionConfig` is what is called when the user clicks the pencil, setting `editCollection` to true and showing the editor menu. When the menu is closed, we want `refreshCollectionSettings` to be called so the `useEffect` calls `refreshCollectionSettings` if `editCollection` changes (it will become false when the editor is closed). This way, the UI for the settings is updates without having to rely on the broadcast events from angular. 

Does that help explain things?

shankari reviewed (approved):
I am fine with merging this as-is, although I would like you to address my comments in a separate cleanup PR

Abby-Wheelis commented on www/js/control/ControlCollectionHelper.tsx:
```diff
@@ -0,0 +1,284 @@
+    return retVal;
+}
+
+const ControlSyncHelper = ({ editVis, setEditVis }) => {
```
You're correct, fixed in [911fd96](https://github.com/Abby-Wheelis/e-mission-phone/commit/911fd9653851ead0d7b21ea63a3d4a7f6c9de875)

Abby-Wheelis commented on www/js/control/ControlSyncHelper.tsx:
```diff
@@ -0,0 +1,284 @@
+        console.log("new config = "+localConfig);
+        try{
+            let set = setConfig(localConfig);
+            //NOTE -- we need to make sure we update these settings in ProfileSettings :) -- getting rid of broadcast handling for migration!!
```
Because the ControlSyncHelper contains the table and editor, the updates immediately show in the UI on save, removing this comment to prevent further confusion.

Abby-Wheelis commented on www/js/control/LogPage.tsx:
```diff
@@ -0,0 +1,153 @@
+                    />
+            </SafeAreaView>
+
+        <AlertBar visible={maxErrorVis} setVisible={setMaxErrorVis} messageKey={""} messageAddition={maxMessage}></AlertBar>
+        <AlertBar visible={logErrorVis} setVisible={setLogErrorVis} messageKey={""} messageAddition={logMessage}></AlertBar>
```
noted for future change: [Next Phase of migrating Profile to React](https://github.com/e-mission/e-mission-docs/issues/946#top)

Abby-Wheelis commented on www/js/control/ControlSyncHelper.tsx:
```diff
@@ -0,0 +1,284 @@
+            <SettingRow textKey="control.end-trip-sync" iconName="sync-alert" action={endForceSync}></SettingRow>
+
+            {/* dataPending */}
+            <Modal visible={dataPendingVis} onDismiss={()=>setDataPendingVis(false)} transparent={true}>
+                <Dialog visible={dataPendingVis} 
+                        onDismiss={()=>setDataPendingVis(false)} 
+                        style={settingStyles.dialog(colors.elevation.level3)}>
+                    <Dialog.Title>{t('data pending for push')}</Dialog.Title>
+                    <Dialog.Actions>
+                        <Button onPress={()=>{
+                            setDataPendingVis(false);
+                            Logger.log("user refused to re-sync")}}>
+                                {t('general-settings.cancel')}
+                        </Button>
+                        <Button onPress={()=>{
+                            setDataPendingVis(false);
+                            forceSync();}}>
+                                {t('general-settings.confirm')}
+                        </Button>
+                    </Dialog.Actions>
+                </Dialog>
+            </Modal>
```
noted for future change: [Next Phase of migrating Profile to React](https://github.com/e-mission/e-mission-docs/issues/946#top)
